File: src/javascript/algorithms/array/1.containsDuplicate/containsDuplicate.js
// Using Set for early exit: TC: O(n), SC: O(n)
function containsDuplicate(nums) {
    const seen = new Set();
    for (const num of nums) {
        if (seen.has(num)) {
            return true; // Duplicate found
        }
        seen.add(num);
    }
    return false; // No duplicates
}

// Using Object for early exit: TC: O(n), SC: O(n)
function containsDuplicateUsingObject(nums) {
    const seen = {};
    for (const num of nums) {
        if (seen[num]) {
            return true; // Duplicate found
        }
        seen[num] = true;
    }
    return false; // No duplicates
}

// Using Set size comparison: TC: O(n), SC: O(n)
function containsDuplicateUsingSize(nums) {
    return new Set(nums).size !== nums.length;
}

// Using sort and iteration: TC: O(n log n), SC: O(1) (modifies input array)
function containsDuplicateUsingSort(nums) {
    nums.sort((a, b) => a - b); // Sort the array
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] === nums[i - 1]) {
            return true; // Duplicate found
        }
    }
    return false; // No duplicates
}

// Using brute-force: TC: O(n^2), SC: O(1) (inefficient for large arrays)
function containsDuplicateUsingBruteforce(nums) {
    for (let i = 0; i < nums.length - 1; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] === nums[j]) {
                return true; // Duplicate found
            }
        }
    }
    return false; // No duplicates
}

// Test cases
console.log("-----Has duplicates----");
const numsWithDuplicates = [8, 6, 4, 2, 6];
console.log(containsDuplicate(numsWithDuplicates));
console.log(containsDuplicateUsingObject(numsWithDuplicates));
console.log(containsDuplicateUsingSize(numsWithDuplicates));
console.log(containsDuplicateUsingSort([...numsWithDuplicates])); // Spread to avoid modifying original
console.log(containsDuplicateUsingBruteforce(numsWithDuplicates));

console.log("-----No duplicates----");
const numsWithoutDuplicates = [1, 3, 5, 7, 9];
console.log(containsDuplicate(numsWithoutDuplicates));
console.log(containsDuplicateUsingObject(numsWithoutDuplicates));
console.log(containsDuplicateUsingSize(numsWithoutDuplicates));
console.log(containsDuplicateUsingSort([...numsWithoutDuplicates])); // Spread to avoid modifying original
console.log(containsDuplicateUsingBruteforce(numsWithoutDuplicates));

File: src/python/algorithms/array/1.containsDuplicate/contains_duplicate.py
"""
Contains Duplicate

Given an integer array nums, return true if any value appears at least twice
in the array, and return false if every element is distinct.

Time Complexity: O(n)
Space Complexity: O(n)
"""


def contains_duplicate(nums):
    """
    Using Set for early exit.
    TC: O(n), SC: O(n)
    """
    seen = set()
    for num in nums:
        if num in seen:
            return True  # Duplicate found
        seen.add(num)
    return False  # No duplicates


def contains_duplicate_using_dict(nums):
    """
    Using dictionary for early exit.
    TC: O(n), SC: O(n)
    """
    seen = {}
    for num in nums:
        if num in seen:
            return True  # Duplicate found
        seen[num] = True
    return False  # No duplicates


def contains_duplicate_using_size(nums):
    """
    Using Set size comparison.
    TC: O(n), SC: O(n)
    """
    return len(set(nums)) != len(nums)


def contains_duplicate_using_sort(nums):
    """
    Using sort and iteration.
    TC: O(n log n), SC: O(1) (modifies input array)
    """
    nums.sort()
    for i in range(1, len(nums)):
        if nums[i] == nums[i - 1]:
            return True  # Duplicate found
    return False  # No duplicates


def contains_duplicate_bruteforce(nums):
    """
    Using brute-force.
    TC: O(n^2), SC: O(1) (inefficient for large arrays)
    """
    for i in range(len(nums) - 1):
        for j in range(i + 1, len(nums)):
            if nums[i] == nums[j]:
                return True  # Duplicate found
    return False  # No duplicates


# Test cases
if __name__ == "__main__":
    test_cases = [
        [1, 2, 3, 1],      # True
        [1, 2, 3, 4],      # False
        [1, 1, 1, 3, 3, 4, 3, 2, 4, 2],  # True
    ]
    
    for nums in test_cases:
        print(f"Input: {nums}")
        print(f"Contains duplicate: {contains_duplicate(nums)}")
        print()
